smack-websocket/src/main/java/org/jivesoftware/smack/websocket/XmppWebsocketTransportModuleDescriptor.java

```diff
@@ -0,0 +1,81 @@
+/**
+ *
+ * Copyright 2020 Aditya Borikar
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smack.websocket;
+
+import java.net.URI;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.jivesoftware.smack.c2s.ModularXmppClientToServerConnectionModule;
+import org.jivesoftware.smack.c2s.ModularXmppClientToServerConnectionModuleDescriptor;
+import org.jivesoftware.smack.c2s.internal.ModularXmppClientToServerConnectionInternal;
+import org.jivesoftware.smack.fsm.StateDescriptor;
+import org.jivesoftware.smack.websocket.XmppWebsocketTransportModule.EstablishingWebsocketConnectionStateDescriptor;
+
+public final class XmppWebsocketTransportModuleDescriptor extends ModularXmppClientToServerConnectionModuleDescriptor {
+    private boolean limitDiscoveryToManualUri;
+    private URI manualUri;
+
+    public XmppWebsocketTransportModuleDescriptor(Builder builder) {
+        this.limitDiscoveryToManualUri = builder.limitDiscoveryToManualUri;
+        this.manualUri = builder.manualUri;
+    }
+
+    public boolean isDiscoveryLimitedToManualUri() {
+        return limitDiscoveryToManualUri;
+    }
+
+    public URI getManualUri() {
+        return manualUri;
+    }
+
+    @Override
+    protected Set<Class<? extends StateDescriptor>> getStateDescriptors() {
+        Set<Class<? extends StateDescriptor>> res = new HashSet<>();
+        res.add(EstablishingWebsocketConnectionStateDescriptor.class);
+        return res;
+    }
+
+    @Override
+    protected ModularXmppClientToServerConnectionModule<? extends ModularXmppClientToServerConnectionModuleDescriptor> constructXmppConnectionModule(
+            ModularXmppClientToServerConnectionInternal connectionInternal) {
+        return new XmppWebsocketTransportModule(this, connectionInternal);
+    }
+
+    public static final class Builder extends ModularXmppClientToServerConnectionModuleDescriptor.Builder {
+        private boolean limitDiscoveryToManualUri = false;
+        private URI manualUri;
+
+        protected Builder(
+                org.jivesoftware.smack.c2s.ModularXmppClientToServerConnectionConfiguration.Builder connectionConfigurationBuilder) {
+            super(connectionConfigurationBuilder);
+        }
+
+        public void setLimitDiscoveryToManualUri(boolean limitDiscoveryToManualUri) {
+            this.limitDiscoveryToManualUri = limitDiscoveryToManualUri;
+        }
+
+        public void setManualUri(URI uri) {
```

Comment:
I think we can safely use `URL` here instead of `URI`. Hence rename the method, and while at it, drop the 'Manual'. Furthermore, add another method `setUrl(URL url, boolean performWebsocketEndpointDiscovery)`, that sets both fields. The rationale is that the user should only be able to disable endpoint discovery if he or she explicitly provides an URL. 

Then
```
Builder setUrl(URL url) {
     return setUrl(url, true)
}
```

Also note that setters in builder *always* return a reference to the current builder (so that you can use them as builder).

Reply:
@Flowdalic Constructing a URL through new URL(endpointString) results into `MalformedURLException: unknown protocol: ws`. 
And hence, we might need to continue with URIs.